uwrt_utils/src/uwrt_can.cpp

```diff
@@ -0,0 +1,129 @@
+  }
+  strcpy(ifr_.ifr_name, interface_name_.c_str());
+  ioctl(socket_handle_, SIOCGIFINDEX, &ifr_);
+  sock_addr_.can_family = AF_CAN;
+  sock_addr_.can_ifindex = ifr_.ifr_ifindex;
+
+  // set software filters on can_id
+  std::vector<struct can_filter> filters;
+  filters.resize(ids.size());
+  for (int i = 0; i < ids.size(); i++) {
+    filters[i].can_id = ids[i];
+    filters[i].can_mask = (CAN_EFF_FLAG | CAN_RTR_FLAG | CAN_SFF_MASK);
+  }
+  setsockopt(socket_handle_, SOL_CAN_RAW, CAN_RAW_FILTER, filters.data(), sizeof(struct can_filter) * filters.size());
+
+  // set read timeout to be very small, so it's not blocking
+  struct timeval timeout {};
+  timeout.tv_usec = 500;
+  setsockopt(socket_handle_, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+
+  // finally bind can socket to can addr
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+  if (bind(socket_handle_, reinterpret_cast<struct sockaddr*>(&sock_addr_), sizeof(sock_addr_)) < 0) {
+    return UWRTCANWrapper::UWRTCANStatus::SOCKET_BIND_FAILED;
+  }
+
+  // start read thread
+  read_thread_running_ = true;
+  read_thread_ = std::thread(&UWRTCANWrapper::readSocketTask, this);
+
+  return UWRTCANWrapper::UWRTCANStatus::STATUS_OK;
+}
+
+// TODO (wraftus) implement heartbeat logic for failover CAN
+void UWRTCANWrapper::readSocketTask() {
+  struct can_frame frame {};
+  int bytes_read;
+
+  while (read_thread_running_) {
+    // keep performing read until buffer is emptied
+    do {
+      bytes_read = recv(socket_handle_, &frame, sizeof(struct can_frame), 0);
+      if (bytes_read == sizeof(struct can_frame)) {
+        if (!recv_map_mtx_.try_lock_for(MUTEX_LOCK_TIMEOUT_)) {
+          break;
+        }
+        recv_map_[frame.can_id] = frame;
+        recv_map_mtx_.unlock();
+      } else {
+        // ruh roh
+      }
+    } while (bytes_read == sizeof(struct can_frame));
+
+    // buffer is emptied, lets wait
+    std::this_thread::sleep_for(thread_sleep_millis_);
+  }
+}
+
+template <class T>
+UWRTCANWrapper::UWRTCANStatus UWRTCANWrapper::getLatestFromID(T& data, canid_t id) {
+  // make sure we have been initialized
+  if (!initialized_) {
+    return UWRTCANWrapper::UWRTCANStatus::NOT_INITED;
+  }
+
+  // check that we have new data to read at specified id
+  std::map<canid_t, struct can_frame>::iterator it;
+  if (!recv_map_mtx_.try_lock_for(MUTEX_LOCK_TIMEOUT_)) {
+    return UWRTCANWrapper::UWRTCANStatus::RECV_MUTEX_TIMEOUT;
+  }
+  it = recv_map_.find(id);
+  if (it == recv_map_.end()) {
+    recv_map_mtx_.unlock();
+    return UWRTCANWrapper::UWRTCANStatus::RECV_UNREGOCNIZED_ID;
+  }
+
+  // read and delete new data at specified id
+  struct can_frame frame = recv_map_[id];
+  recv_map_.erase(it);
+  recv_map_mtx_.unlock();
+
+  // extract data from frame
+  if (frame.can_dlc != sizeof(data)) {
+    return UWRTCANWrapper::UWRTCANStatus::RECV_SIZE_MISMATCH;
+  }
+  memcpy(&data, frame.data, sizeof(T));
```

Comment:
this only works if endianess matches. should have check for system's endianess.

Reply:
yea wasn't sure if we should do endian changes here or leave it up to user, but i guess it's better to take it in as a flag and do it ourselves to avoid seven change endian functions

Reply:
maybe take in a flag in constructor to indicate can frame's endianess and then get the machine's endianess automatically

Reply:
https://en.cppreference.com/w/cpp/types/endian

Reply:
^^ this is only in c++20

Reply:
it gives u the sample implementation so you can just have that as a helper function in your package

Reply:
can be smth like this https://stackoverflow.com/a/38141476

Reply:
yeah i just used those `#defines` to figure out when to swap endianness
another thing is that if we send over structs endianness might get a bit wonky with the template type I think

Reply:
what do you mean by "we send over structs"

Reply:
like if we ever send more than one value over can, like an int and a float for example, we'd need to correct the endianness of those two values separately, right?

Reply:
yeah if the type was a struct, it would not work. T can only be primatives under 8 bytes long. is there a way you could check for that? maybe explicitly instantiating the templates we say are allowed? don't think we'd ever send a struct. if anything, if we want to pack multiple values, we can address that later.